pyclibrary/c_parser.py

```diff
@@ -407,15 +153,16 @@ class CParser(object):
     def __init__(self, files=None, copy_from=None, replace=None,
                  process_all=True, cache=None, **kwargs):
 
+        self.clib_intf = c_model.CLibInterface()
+        if copy_from is not None:
+            self.clib_intf.include(copy_from)
+        self.macro_vals = {}
+
         if not self._init:
             logger.warning('Automatic initialisation based on OS detection')
             from .init import auto_init
             auto_init()
 
-        # Holds all definitions
-        self.defs = {}
-        # Holds definitions grouped by the file they came from
-        self.file_defs = {}
         # Description of the struct packing rules as defined by #pragma pack
```

Comment:
I don't see any problem here:

The preprocessor gets a function (i.e.```process_file()```) that gets a string (the unprocessed C source code) and a clibinterface object (where the macro defs shall be added into) and returns a string (the C source code, where the macros are resolved).

the parser class gets a clibinterface (where all parser specific defs thall be added into) and a string (the C source code).

Furthermore we could provide a (global) convinience function, where you could provide a list of files objects / file names. This function will:
 * create a CLibInterface
 * run the preprocessor for every source file
 * run the parser for every preprocessed source file

Or did I miss something / did not get your point?